src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,148 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
+		idx_t offset = start_idx;
+		for (idx_t i = start_idx + 1; i <= end_idx; i++) {
+			if (keys[i - 1]->data[depth] != keys[i]->data[depth]) {
+				BuildARTOffsets entry(offset, i - 1, keys[i - 1]->data[depth]);
+				offsets.push_back(entry);
+				offset = i;
+			}
+		}
+		BuildARTOffsets entry(offset, end_idx, keys[end_idx]->data[depth]);
+		offsets.push_back(entry);
+
+		if (offsets.size() < 5) {
+			node = new Node4();
+		} else if (offsets.size() < 17) {
+			node = new Node16();
+		} else if (offsets.size() < 49) {
+			node = new Node48();
+		} else {
+			node = new Node256();
+		}
+
+		auto prefix_length = depth - prefix_start;
+		node->prefix = Prefix(*keys[start_idx], prefix_start, prefix_length);
+
+		// recurse on each offset
+		for (idx_t i = 0; i < offsets.size(); i++) {
+			Node *child_node = nullptr;
+			Build(keys, row_ids, child_node, offsets[i].start, offsets[i].end, depth + 1);
+			Node::InsertChildNode(node, offsets[i].byte, child_node);
+		}
+	}
+}
+
+void ART::BuildAndMerge(IndexLock &lock, PayloadScanner &scanner, Allocator &allocator) {
+
+	auto payload_types = logical_types;
+	payload_types.emplace_back(LogicalType::ROW_TYPE);
+
+	auto skipped_all_nulls = false;
+	auto temp_art = make_unique<ART>(this->column_ids, this->unbound_expressions, this->constraint_type, this->db);
+	for (;;) {
+		DataChunk ordered_chunk;
+		ordered_chunk.Initialize(allocator, payload_types);
+		ordered_chunk.SetCardinality(0);
+		scanner.Scan(ordered_chunk);
+		if (ordered_chunk.size() == 0) {
+			break;
+		}
+
+		// get the key chunk and the row_ids vector
+		DataChunk chunk_row_ids;
+		ordered_chunk.Split(chunk_row_ids, ordered_chunk.ColumnCount() - 1);
+		auto &row_ids = chunk_row_ids.data[0];
+
+		D_ASSERT(row_ids.GetType().InternalType() == ROW_TYPE);
+		D_ASSERT(logical_types[0] == ordered_chunk.data[0].GetType());
+
+		// generate the keys for the given input
+		vector<unique_ptr<Key>> keys;
+		GenerateKeys(ordered_chunk, keys);
+
+		idx_t start_idx = 0;
+		if (!skipped_all_nulls) {
+			for (idx_t i = 0; i < keys.size(); i++) {
+				if (keys[i]) {
+					start_idx = i;
+					skipped_all_nulls = true;
+					break;
+				}
+			}
+		}
+
+		if (start_idx != 0 && IsPrimary()) {
+			throw ConstraintException("NULLs in new data violate the primary key constraint of the index");
+		}
+
+		if (!skipped_all_nulls) {
+			if (IsPrimary()) {
+				// chunk consists only of NULLs
+				throw ConstraintException("NULLs in new data violate the primary key constraint of the index");
+			}
+			continue;
+		}
+
+		// prepare the row_ids
+		row_ids.Flatten(ordered_chunk.size());
+		auto row_identifiers = FlatVector::GetData<row_t>(row_ids);
+
+		// build the ART of this chunk
+		auto art = make_unique<ART>(this->column_ids, this->unbound_expressions, this->constraint_type, this->db);
+		Build(keys, row_identifiers, art->tree, start_idx, ordered_chunk.size() - 1, 0);
+
+		// merge art into temp_art
+		ART::Merge(temp_art.get(), art.get());
```

Comment:
Maybe it makes sense to merge both exceptions and have an if clause for the 331-339 lines.